pkg/rhcos/release/release.go

```diff
@@ -0,0 +1,180 @@
+// Package release extracts the RHCOS build version from an
+// OpenShift release image.
+package release
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"io"
+	"io/ioutil"
+
+	"github.com/containers/image/docker"
+	"github.com/containers/image/image"
+	"github.com/containers/image/pkg/blobinfocache"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/transports"
+	"github.com/containers/image/types"
+	"github.com/docker/distribution/reference"
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+// RHCOSBuild extracts the RHCOS build version from an OpenShift
+// release image.
+func RHCOSBuild(ctx context.Context, releaseImagePullSpec string, pullSecret []byte) (string, error) {
+	imageReferencesBytes, err := imageReferences(ctx, releaseImagePullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+
+	pullSpec, err := osContentPullSpec(imageReferencesBytes)
+	if err != nil {
+		return "", err
+	}
+
+	imageSource, image, err := pull(ctx, pullSpec, pullSecret)
+	if err != nil {
+		return "", err
+	}
+	defer imageSource.Close()
+
+	inspect, err := image.Inspect(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	version, ok := inspect.Labels["version"]
+	if !ok {
+		return "", errors.Errorf("%s has no version annotation", pullSpec)
+	}
+
+	return version, nil
+}
+
+// pull pulls an image.  If the pull secret includes credentials for
+// the pull-spec authority, those are used to authenticate requests.
+// The signature on the returned image, if any, is validated agaist
+// the signature policy from /etc/containers/policy.json.
+func pull(ctx context.Context, pullSpec string, pullSecret []byte) (types.ImageSource, types.Image, error) {
```

Comment:
> (FWIW, being pedantic, this does not “pull” the (entirety of) an image; it basically fetches only the config and manifest, not the layer blobs.)

Yeah, but `pull` is a well-established term in this context, and I didn't want to rename just because we're likely to bail out from what folks would usually call a "pull" before we'd worked through all the images.  I'm open to renames if folks want to push for an alternative.

Reply:
Naming within this project is of course up to this project :)